### ISSUE TITLE
Builder Component "Array to string conversion"

### ISSUE THREAD

**Issue by yabasha** [rainlab/builder-plugin#162]
Hi,

When I want to add Builder component on a page I get the following error message:

"Array to string conversion" on line 36 of /home/vagrant/Code/octobercms/vendor/laravel/framework/src/Illuminate/Database/Schema/MySqlBuilder.php

and when I change the Model class to one of the models I already have following error popup:
Invalid Model Class Name.

Please advice,

![image](https://cloud.githubusercontent.com/assets/932321/26302926/69a76faa-3eee-11e7-858e-f8375e594afa.png)

![image](https://cloud.githubusercontent.com/assets/932321/26302850/40d90962-3eee-11e7-8f6b-9cf2a9178417.png)

**Comment by fladder70** [comment]
Hi - I've got exactly the same. I've tried going through the code to identify the problem but no luck so far. Anyone?

**Comment by yabasha** [comment]
I downgraded to earlier version (1.0.11) where the plugin works correctly.

**Comment by LukeTowers** [comment]
Somebody post a full stack trace please.

**Comment by tobbi** [comment]
Stack trace: https://pastebin.com/UvZNVEfe

**Comment by LukeTowers** [comment]
And @tobbi could you post detailed replication instructions?

**Comment by vkscorpion1986** [comment]
I'm getting the same issue as "yabasha" posted on 22 May. Please suggest some solution.

**Comment by LukeTowers** [comment]
@vkscorpion1986 post detailed replication instructions. 
The amount of time spent asking for help is directly related to the amount of time spent putting in effort to answer it. If you want help and / or guidance, then invest as much time as possible in making questions that are detailed and easy to understand. Otherwise, nobody is going to invest their time helping with a question that wasn't even worth the askers time to ask correctly.

**Comment by tleish** [comment]
# "Array to string conversion" error

I ran into this same issues stated above.  

The first error is sometimes tricky to reproduce.  Sometimes you must repeat the following steps:

1. Go to CMS > Pages > [any page]
2. Navigate to Components
3. Drag new Builder > "Record List"  onto the page.
4. Click the new component to show options.

The error is then thrown.

> "Array to string conversion" on line 30 of /vendor/laravel/framework/src/Illuminate/Database/Schema/MySqlBuilder.php

I believe it stems from the following Builder Code

### Illuminate\Database\Schema\MySqlBuilder
This is the line of code where the error is thrown.  `$table` is expected to be a string, but in the event of an error, it's an empty array.
```php
    public function getColumnListing($table)
    {
        $table = $this->connection->getTablePrefix().$table;
        // ...
    }
```

#### RainLab\Builder\Classes\ModelModel
This function calls Schema::getColumnListing($tableName) from the above class, and sometimes passes in `$tableName = []`;
```php
    public static function getModelFields($pluginCodeObj, $modelClassName)
    {
        $tableName = self::getTableNameFromModelClass($pluginCodeObj, $modelClassName);

        // Currently we return only table columns,
        // but eventually we might want to return relations as well.

        return Schema::getColumnListing($tableName);
    }
```

In this particular case:

* `$modelClassName = 'Settings'`
* Which then creates: `$modelFilePath = '/local/path/plugins/rainlab/builder/models/Settings.php'`
* Which then hits the following condition: 

#### RainLab\Builder\Classes\ModelModel
```php
    private static function getTableNameFromModelClass($pluginCodeObj, $modelClassName)
       //...
        if (!$modelInfo || !isset($modelInfo['table'])) {
            return [];
        }
```

I believe the problem is that `getTableNameFromModelClass` method will return an array instead of a string in the event of an error.

The following changes removed the error from my tests, but I did not research to see if this caused issues elsewhere.

```diff
    private static function getTableNameFromModelClass($pluginCodeObj, $modelClassName)
    {
        if (!self::validateModelClassName($modelClassName)) {
            throw new SystemException('Invalid model class name: '.$modelClassName);
        }

        $modelsDirectoryPath = File::symbolizePath($pluginCodeObj->toPluginDirectoryPath().'/models');
        if (!File::isDirectory($modelsDirectoryPath)) {
-            return [];
+            return '';
        }

        $modelFilePath = $modelsDirectoryPath.'/'.$modelClassName.'.php';
        if (!File::isFile($modelFilePath)) {
-            return [];
+            return '';
        }

        $parser = new ModelFileParser();
        $modelInfo = $parser->extractModelInfoFromSource(File::get($modelFilePath));
        if (!$modelInfo || !isset($modelInfo['table'])) {
-            return [];
+            return '';
        }

        return $modelInfo['table'];
    }
```

**Comment by tleish** [comment]
Note: issue for **Invalid model class name** is a separate issue and addressed in #202

**Comment by LukeTowers** [comment]
https://github.com/rainlab/builder-plugin/commit/7efae36016f980a6383f6cb308893311efbf1618 fixes the invalid class name, @tleish made #203 to fix the array to string conversion; just waiting on the tests for that.
